batteries/difftext/myersdiff.luau

```diff
@@ -0,0 +1,185 @@
+local deque = require("@batteries/collections/deque")
+
+type EditGraphNode = {
+	x: number,
+	y: number,
+	prev: EditGraphNode?,
+}
+
+local function editGraphNode(x: number, y: number, prev: EditGraphNode?): EditGraphNode
+	return {
+		x = x,
+		y = y,
+		prev = prev,
+	}
+end
+
+-- true if a[x+1] == b[y+1]; indicates an EQUAL operation (no diff) is valid
+local function hasDiagonalEdge(node: EditGraphNode, a: string | { string }, b: string | { string }): boolean
+	local x, y = node.x, node.y
+	if typeof(a) == "table" and typeof(b) == "table" then
+		return x < #a and y < #b and a[x + 1] == b[y + 1]
+	end
+	a = a :: string
+	b = b :: string
+	return x < #a and y < #b and a:sub(x + 1, x + 1) == b:sub(y + 1, y + 1)
+end
+
+export type DiffOperation = {
+	key: "EQUAL" | "ADD" | "DELETE",
+	text: string,
+	position: number?,
+}
+
+export type Diff = { DiffOperation }
+
+local function edgeToDiff(curNode: EditGraphNode, a: string | { string }, b: string | { string }): DiffOperation
+	-- maps stepping from curNode.prev -> curNode to a diff operation
+	-- (n, n) -> (n+1, n) == DELETION of a[n]
+	-- (n, n) -> (n, n+1) == ADDITION of b[n]
+	-- (n, n) -> (n+1, n+1) == EQUAL (no diff)
+	local prev = curNode.prev :: EditGraphNode
```

Comment:
can we return `nil` if `currNode.prev == nil` instead? this isn't making the fullest use of the type solver